lib/Data/Chronicle/Writer.pm

```diff
@@ -128,38 +127,119 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
 
-    die "Recorded date is undefined" unless $rec_date;
-    die "Recorded date is not a Date::Utility object" if ref $rec_date ne 'Date::Utility';
-    die "Cannot store undefined values in Chronicle!" unless defined $value;
-    die "You can only store hash-ref or array-ref in Chronicle!" unless (ref $value eq 'ARRAY' or ref $value eq 'HASH');
+    $self->mset([[$category, $name, $value]], $rec_date, $archive, $suppress_pub, $ttl);
+
+    return 1;
+}
+
+=head2 mset
 
-    $value = JSON::MaybeXS->new->encode($value);
+Example:
+
+    $chronicle_writer->mset([["category1", "name1", $value1], ["category2, "name2", $value2], ...]);
+
+Store a piece of data "value1" under key "category1::name1", "category2::name2", etc in Pg and Redis. Will
+publish "category1::name1", "category2::name2", etc in Redis if C<publish_on_set> is true.
+
+=cut
+
+sub mset {
+    my $self         = shift;
+    my $entries      = shift;
+    my $rec_date     = shift;
+    my $archive      = shift // 1;
+    my $suppress_pub = shift // 0;
+    my $ttl          = shift // $self->ttl;
+
+    $self->_validate_value($_->[2]) foreach @$entries;
+    $self->_validate_rec_date($rec_date);
 
-    my $key    = $category . '::' . $name;
     my $writer = $self->cache_writer;
 
     # publish & set in transaction
     $writer->multi;
-    my $encoded = encode_utf8($value);
-    $writer->publish($key, $encoded) if $self->publish_on_set;
-    $writer->set(
-        $key => $encoded,
-        $ttl ? ('EX' => $ttl) : ());
-    $writer->exec;
+    foreach my $entry (@$entries) {
+        my $category = $entry->[0];
+        my $name     = $entry->[1];
+        my $value    = $entry->[2];
+
+        my $key = $self->_generate_key($category, $name);
+
+        my $encoded = encode_json_utf8($value);
+        $writer->publish($key, $encoded) if $self->publish_on_set && !$suppress_pub;
+        $writer->set(
+            $key => $encoded,
+            $ttl ? ('EX' => $ttl) : ());
 
-    $self->_archive($category, $name, $value, $rec_date) if $archive and $self->dbic;
+        $self->_archive($category, $name, $encoded, $rec_date) if $archive and $self->dbic;
+    }
+    $writer->exec;
 
     return 1;
 }
 
+=head2 subscribe
+
+Example:
+
+    $chronicle_writer->subscribe("category1", "name1", $code_ref);
+
+=cut
+
+sub subscribe {
+    my ($self, $category, $name, $subref) = @_;
+    die 'publish_on_set must be enabled to subscribe' unless $self->publish_on_set;
+    die 'Subscription requires a coderef' if ref $subref ne 'CODE';
+
+    my $key = $self->_generate_key($category, $name);
+    return $self->cache_writer->subscribe($key, $subref);
+}
+
+=head2 unsubscribe
+
+Example:
+
+    $chronicle_writer->unsubscribe("category1", "name1", $code_ref);
+
+=cut
+
+sub unsubscribe {
+    my ($self, $category, $name, $subref) = @_;
+    die 'publish_on_set must be enabled to unsubscribe' unless $self->publish_on_set;
+    die 'Unsubscription requires a coderef' if ref $subref ne 'CODE';
```

Comment:
Any reason you have to pass the same coderef for unsubscribe? That makes it difficult to easily perform subscription using code like

    $writer->subscribe(category => name => sub { ... });

because now you'll have to have saved that particular coderef to pass it later for unsubscribe.

Reply:
I think I made an error here by looking at the wrong documentation, I will go back and change this. Thank you for also bringing to my attention the issue of Redis connection in the previous comment.